usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -908,9 +918,13 @@ zil_lwb_write_init(zilog_t *zilog, lwb_t *lwb)
 		    ZIO_FLAG_CANFAIL);
 	}
 	if (lwb->lwb_zio == NULL) {
+		if (zilog->zl_cur_used <= zil_slog_limit || !lwb->lwb_slog)
+			prio = ZIO_PRIORITY_SYNC_WRITE;
+		else
+			prio = ZIO_PRIORITY_ASYNC_WRITE;
```

Comment:
Async writes are processed in sorted order (by offset), so a steady influx of async writes can starve out some async writes for a long time (imagine one thread that issues a write to the end of the disk, while another thread is writing to every offset starting at the beginning -- the first thread's write will not be processed until the second thread has written the whole disk).  This may not happen commonly for SLOG devices, but it makes it pretty difficult to reason about the performance.